picasso/src/test/java/com/squareup/picasso3/NetworkRequestHandlerTest.kt

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (C) 2013 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.picasso3
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.picasso3.RequestHandler.Result
+import com.squareup.picasso3.TestUtils.PremadeCall
+import com.squareup.picasso3.TestUtils.URI_1
+import com.squareup.picasso3.TestUtils.URI_KEY_1
+import com.squareup.picasso3.TestUtils.mockNetworkInfo
+import okhttp3.CacheControl
+import okhttp3.MediaType
+import okhttp3.Protocol.HTTP_1_1
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.ResponseBody
+import okhttp3.ResponseBody.Companion.toResponseBody
+import okio.Buffer
+import okio.BufferedSource
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers.anyLong
+import org.mockito.Mock
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.never
+import org.mockito.Mockito.verify
+import org.mockito.MockitoAnnotations.initMocks
+import org.robolectric.RobolectricTestRunner
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.TimeUnit.SECONDS
+import java.util.concurrent.atomic.AtomicBoolean
+
+@RunWith(RobolectricTestRunner::class)
+class NetworkRequestHandlerTest {
+  private val responses = LinkedBlockingDeque<Response>()
+  private val requests = LinkedBlockingDeque<Request>()
+
+  @Mock internal lateinit var picasso: Picasso
+  @Mock internal lateinit var dispatcher: Dispatcher
+  private lateinit var networkHandler: NetworkRequestHandler
+
+  @Before fun setUp() {
+    initMocks(this)
+    networkHandler = NetworkRequestHandler { request ->
+      requests.add(request)
+      try {
+        PremadeCall(request, responses.takeFirst())
+      } catch (e: InterruptedException) {
+        throw AssertionError(e)
+      }
+    }
+  }
+
+  @Test @Throws(Exception::class)
```

Comment:
nit: are these `@Throws` annotations useful? In Java I believe they were needed to satisfy the compiler, but Kotlin doesn't care. Or are they worth keeping for documentation purposes?

Reply:
good point! i'll clean that up along with @JvmStatic, etc. once everything is converted.